cpp/src_prims/sparse/distance/coo_spmv.cuh

```diff
@@ -0,0 +1,306 @@
+  // encountered by each thread
+  for (int i = tid; i < active_chunk_size; i += blockDim.x) {
+    value_idx ind_next = ind + blockDim.x;
+    value_idx next_row_b = -1;
+
+    if (i + blockDim.x < active_chunk_size) next_row_b = rowsB[ind_next];
+
+    if (next_row_b != cur_row_b) {
+      unsigned int peer_group = get_peer_group(cur_row_b);
+      bool is_leader = get_lowest_peer(peer_group) == lane_id;
+
+      value_t v = warp_red.HeadSegmentedReduce(c, is_leader, accum_func);
+
+      // thread with lowest lane id among peers writes out
+      if (is_leader && v != 0.0) {
+        value_idx idx =
+          !rev ? cur_row_a * n + cur_row_b : cur_row_b * m + cur_row_a;
+        write_func(out + idx, v);
+      }
+      c = 0.0;
+    }
+
+    if (next_row_b != -1) {
+      ind = ind_next;
+      value_t a_col = A[indicesB[ind]];
+      if (!rev || a_col == 0.0)
+        c = accum_func(c, reduce_func(a_col, dataB[ind]));
+      cur_row_b = next_row_b;
+    }
+  }
+}
+
+/**
+ * Computes the maximum number of columns that can be stored
+ * in shared memory in dense form with the given block size
+ * and precision.
+ * @return the maximum number of columns that can be stored in smem
+ */
+template <typename value_idx, typename value_t, int tpb = 1024>
+inline int max_cols_per_block() {
+  // compute max shared mem to use
+  return (raft::getSharedMemPerBlock() - (2 * sizeof(value_idx)) -
+          ((tpb / raft::warp_size()) * sizeof(value_t))) /
+         sizeof(value_t);
+}
+
+/**
+ * Performs generalized sparse-matrix-sparse-matrix multiplication via a
+ * sparse-matrix-sparse-vector layout. Each vector of A is loaded
+ * into shared memory in dense form and the non-zeros of B
+ * load balanced across the threads of each block.
+ * @tparam value_idx index type
+ * @tparam value_t value type
+ * @tparam threads_per_block block size
+ * @tparam chunk_size number of nonzeros of B to process for each row of A
+ * @tparam reduce_f semiring product() function
+ * @tparam accum_f semiring sum() function
+ * @tparam write_f atomic semiring sum() function
+ * @param[out] out_dists dense array of out distances of size m * n
+ * @param[in] config_ distance config object
+ * @param[in] coo_rows_b coo row array for B
+ * @param[in] reduce_func semiring product() function
+ * @param[in] accum_func semiring sum() function
+ * @param[in] write_func atomic semiring sum() function
+ */
+template <typename value_idx, typename value_t, int threads_per_block = 1024,
+          int chunk_size = 500000, typename reduce_f, typename accum_f,
+          typename write_f>
+inline void balanced_coo_pairwise_generalized_spmv(
+  value_t *out_dists, const distances_config_t<value_idx, value_t> &config_,
+  value_idx *coo_rows_b, reduce_f reduce_func, accum_f accum_func,
+  write_f write_func) {
+  CUDA_CHECK(cudaMemsetAsync(
+    out_dists, 0, config_.a_nrows * config_.b_nrows * sizeof(value_t),
+    config_.stream));
+  int n_warps_per_row =
```

Comment:
`n_blocks_per_row` to be consistent with the naming in the kernel above, as well as correctly describe the use of this variable?

Reply:
Ah good catch. Originally, this was warps per row because I had the block size set to a warp.